pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/tenant/DefaultTenantRebalancer.java

```diff
@@ -49,6 +56,80 @@ public DefaultTenantRebalancer(PinotHelixResourceManager pinotHelixResourceManag
+        dryRunResults.put(table, new RebalanceResult(null, RebalanceResult.Status.FAILED, exception.getMessage(),
+            null, null, null, null, null));
+      }
+    });
+    if (config.isDryRun()) {
+      return new TenantRebalanceResult(null, dryRunResults, config.isVerboseResult());
+    }
+
+    String tenantRebalanceJobId = createUniqueRebalanceJobIdentifier();
+    TenantRebalanceObserver observer = new ZkBasedTenantRebalanceObserver(tenantRebalanceJobId, config.getTenantName(),
+        tables, _pinotHelixResourceManager);
+    observer.onTrigger(TenantRebalanceObserver.Trigger.START_TRIGGER, null, null);
+    ConcurrentLinkedQueue<String> parallelQueue = createTableQueue(config, dryRunResults);
+    // ensure atleast 1 thread is created to run the sequential table rebalance operations
+    int parallelism = Math.max(config.getDegreeOfParallelism(), 1);
+    try {
+      for (int i = 0; i < parallelism; i++) {
+        _executorService.submit(() -> {
+          while (true) {
+            String table = parallelQueue.poll();
+            if (table == null) {
+              break;
+            }
+            RebalanceConfig rebalanceConfig = RebalanceConfig.copy(config);
+            rebalanceConfig.setDryRun(false);
+            if (dryRunResults.get(table)
+                .getRebalanceSummaryResult()
+                .getSegmentInfo()
+                .getReplicationFactor()
+                .getExpectedValueAfterRebalance() == 1) {
+              rebalanceConfig.setMinAvailableReplicas(0);
+            }
+            rebalanceTable(table, rebalanceConfig, dryRunResults.get(table).getJobId(), observer);
+          }
+          observer.onSuccess(String.format("Successfully rebalanced tenant %s.", config.getTenantName()));
+        });
+      }
+    } catch (Exception exception) {
+      observer.onError(String.format("Failed to rebalance the tenant %s. Cause: %s", config.getTenantName(),
+          exception.getMessage()));
+    }
+
+    // Prepare tenant rebalance result to return
+    Map<String, RebalanceResult> rebalanceResults = new HashMap<>();
+    for (String table : dryRunResults.keySet()) {
+      RebalanceResult result = dryRunResults.get(table);
+      if (result.getStatus() == RebalanceResult.Status.DONE) {
+        rebalanceResults.put(table, new RebalanceResult(result.getJobId(), RebalanceResult.Status.IN_PROGRESS,
+            "In progress, check controller task status for the", result.getInstanceAssignment(),
+            result.getTierInstanceAssignment(), result.getSegmentAssignment(), result.getPreChecksResult(),
+            result.getRebalanceSummaryResult()));
+      } else {
+        rebalanceResults.put(table, dryRunResults.get(table));
+      }
+    }
+    return new TenantRebalanceResult(tenantRebalanceJobId, rebalanceResults, config.isVerboseResult());
+  }
+
+  public TenantRebalanceResult rebalanceWithParallelAndSequential(TenantRebalanceConfig config) {
```

Comment:
Let's add comments about this - explain this is the older way to be deprecated etc